src/packet.c

```diff
@@ -19,13 +20,34 @@ PacketKind *decode(uint8_t *buf, size_t len, Error *err) {
 		return NULL;
 	}
 
-	PacketHeader header = decode_packet_header(buf);
+	Packet *packet = malloc(sizeof(Packet));
+	memset(packet, 0, sizeof(*packet));
+	packet->header = decode_packet_header(buf);
 
-	if(header.flags & PACKET_FLAG_CONTROL) {
-		return (PacketKind *)decode_control(&buf[3], len - 3, header, err);
+	if(packet->header.flags & PACKET_FLAG_CONTROL) {
+		packet->kind = PACKET_CONTROL;
+		packet->control = decode_control(&buf[3], len - 3, &packet->header, err);
 	} else {
-		return (PacketKind *)decode_normal(&buf[3], len - 3, header, err);
+		packet->kind = PACKET_NORMAL;
+		Error chunk_error = fetch_chunks(&buf[3], len - 3, packet);
+		if(chunk_error != ERR_NONE) {
+			if(err) {
+				*err = chunk_error;
+			}
+			free(packet);
```

Comment:
Shouldn't `free_packet` be called here? If error happened on 5th message, it would free the packet pointer but none of the messages